substrate/frame/support/src/traits/hooks.rs

```diff
@@ -99,6 +99,22 @@ pub trait OnGenesis {
 	fn on_genesis() {}
 }
 
+/// Implemented by pallets, allows defining logic to run prior to any [`OnRuntimeUpgrade`] logic.
+///
+/// This was introduced to allow pallets to initialize their on-chain storage version in the case
+/// that they are added to the runtime after genesis.
+///
+/// This hook is intended to be used internally in FRAME and not be exposed to FRAME developers.
+///
+/// It is defined as a seperate trait from [`OnRuntimeUpgrade`] precisely to not pollute the public
+/// API.
+pub trait BeforeAllRuntimeMigrations {
```

Comment:
in principle we can `shield` this to prevent anyone other than `FRAME` macros to implement it.

Reply:
That sounds good, are you able to point me in the direction for how that could be achieved?

Reply:
You can not do this. There is no way to only allow our macros to implement a trait.

Reply:
I was hinting at the trait sealing technique but indeed it doesn't properly work here, because the trait that I am intending to seal is implemented at the end of the day by `Pallet` which is in a different crate. So, by definition, that other crate must have access to the sealed trait as well. 

We could still semi-seal it by putting the `trait Sealed {}` in `mod __private` and give it some name like `trait DoNotImplement {}`, but yeah it won't be proper sealing.

Reply:
`trait ShittySealingPleaseSerDoNotImplementPlease {}` lol :P